packages/neo-one-smart-contract-compiler/src/__tests__/compile/expression/BinaryExpressionCompiler.test.ts

```diff
@@ -351,45 +439,370 @@ if (true && false) {
       if ((undefined || 0) !== 0) {
         throw 'Failure';
       }
     `);
   });
 
-  test('null == undefined', async () => {
+  test('! (true == false) [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (true == false) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('! (undefined == "a") [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (undefined == "a") {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('! (0 == null) [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (0 == null) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test('0 == false [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(0 == false)) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test('null == undefined [EqualsEqualsToken]', async () => {
     await helpers.executeString(`
       if (!(null == undefined)) {
         throw 'Failure';
       }
     `);
   });
 
+  test.skip('!(true == undefined) [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (true == undefined) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('! (false == "a") [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(false == "a")) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('!(null == true) [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (null == true) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('! (null == false) [EqualsEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (null == false) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('true != null [ExclamationEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(true != null)) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('true != undefined [ExclamationEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(true != undefined)) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test('4 != 3 [ExclamationEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(4 != 3)) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test('0 != 3 [ExclamationEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(0 != 3)) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('0 != "a" [ExclamationEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(0 != "a")) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test.skip('undefined != "a" [ExclamationEqualsToken]', async () => {
+    await helpers.executeString(`
+      if (!(undefined != "a")) {
+        throw 'Failure';
+      }
+    `);
+  });
+
+  test('(6 & 5) === 4 [AmpersandToken]', async () => {
+    await helpers.executeString(`
+      if (!(6 & 5) === 4) {
```

Comment:
I think you need to add parentheses: `!((6 & 5) === 4)`. Makes it clearer anyways. Or just use `!==`.